Add questions to forms

tfidfwastaken opened the pull request:
Closes [sc-167].

tfidfwastaken committed:
migrations: create question table

tfidfwastaken committed:
specs: add specs for question

tfidfwastaken committed:
question: add questions to forms in form builder

Allow users to add a question to created form

Currently, questions cannot be deleted.

tfidfwastaken committed:
tests: add factories for question data

tfidfwastaken committed:
tests: test question addition handlers and functions

yogirajh007 commented on test/bugle_forms/handlers/question_test.clj:
```diff
@@ -0,0 +1,33 @@
+          response (sut/add-question request)]
+      (is (= 303 (:status response)))
+      (is (.startsWith (get (:headers response) "Location") "/form-builder/"))
+      (is (not-empty (sql/find-by-keys db/datasource :question
```
Tests for failure could be added

tfidfwastaken replied:
The only failure states that I can think of for this handler is:
- Malformed parameters (responsibility of the validation middleware which has undergone testing)
- Some unexpected database error (responsibility of error handling middleware which has undergone testing)

Is there any other negative case I'm missing? Should I retest these things here as well?

tfidfwastaken replied:
I discussed this with @prabhanshuguptagit since you were in a meeting. His opinion was that the two things do not need to be tested in the handler layer. The one benefit of re-testing database exceptions and malformed requests again in the domain functions is to verify if the correct middleware has been applied. Even then, that is better tested in the routing layer—I have opened a ticket for that.

Is there anything specific to the domain of "adding questions" that you feel is still missing? If not, does it look okay to merge?

tfidfwastaken replied:
> Even then, that is better tested in the routing layer

I have done these. I'm merging now. Do let me know if any change is required.

tfidfwastaken committed:
tests: check access control definitions

We add a route test for ensuring that we have correctly defined access
control definitions. This in combination with the middleware test
exhaustively covers the whole flow of the app handlers.

neenaoffline commented on resources/migrations/20220317055918-create-question.up.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS question (
```
Having an IF NOT EXISTS in these CREATE commands means that the outcome of successfully running these migrations will not guarantee that the `question` table exists with these columns, just that it exists. Additionally, if the `question` table already exists, that probably isn't something we want to silently accept and move on. We should prefer that the migration fails in that situation so that we know that a table with the same name somehow already exists.

tfidfwastaken replied:
I'll fix it when I start the next story. All tables have this issue. Btw, how do we fix this without editing the migration file? In this case I can get away with it, but I wonder how to simulate this in a more "real" project.

One strategy I can think of: repeat all the migrations I have, but without the conditionals.

neenaoffline replied:
Don't change the migrations merged to master.

IIRC, the migration system will recognize that a previous migration has been edited and throw up. So it isn't something one can do anyway.

neenaoffline commented on resources/migrations/20220317055918-create-question.down.sql:
```diff
@@ -0,0 +1 @@
+DROP TABLE IF EXISTS question;
```
Ditto the comment about IF NOT EXISTS in the create migration. Although, it is much less harmful here.
